crates/phactory/src/system/mod.rs

```diff
@@ -514,6 +522,46 @@ impl<Platform: pal::Platform> System<Platform> {
         self.get_system_message_handler(&cluster_id)
     }
 
+    // The IdentityKey is considered valid in two situations:
+    //
+    // 1. It's generated by pRuntime thus is safe;
+    // 2. It's handovered, but we find out that it was successfully registered as a worker on-chain;
+    pub fn validated_identity_key(&self) -> bool {
+        self.trusted_identity_key || self.worker_state.registered
+    }
+
+    pub fn get_worker_key_challenge(
+        &mut self,
+        dev_mode: bool,
+    ) -> HandoverChallenge<chain::BlockNumber> {
+        let payload = HandoverChallengePayload {
+            block_number: self.block_number,
+            now: self.now_ms,
+            dev_mode,
+            nonce: crate::generate_random_info(),
+        };
+        let signature = self.identity_key.sign_data(&payload.encode());
+        let challenge = HandoverChallenge { payload, signature };
+        self.last_challenge = Some(challenge.clone());
+        challenge
+    }
+
+    pub fn verify_worker_key_challenge(
+        &mut self,
+        challenge: &HandoverChallenge<chain::BlockNumber>,
+    ) -> bool {
+        if self.last_challenge.is_none()
+            || *self.last_challenge.as_ref().expect("checked; qed;") != *challenge
```

Comment:
```suggestion
        if self.last_challenge.as_ref() != Some(challenge)
```

Reply:
This contains some legacy code, removed now.